File: generated-code/codecraft/csharp/TcpReadWrite/Model/PlayerView.cs
namespace TransGenTest.Model
{
    /// <summary>
    /// Information available to the player
    /// </summary>
    public struct PlayerView
    {
        /// <summary>
        /// Your player's ID
        /// </summary>
        public int MyId { get; set; }
        /// <summary>
        /// Size of the map
        /// </summary>
        public int MapSize { get; set; }
        /// <summary>
        /// Whether fog of war is enabled
        /// </summary>
        public bool FogOfWar { get; set; }
        /// <summary>
        /// Entity properties for each entity type
        /// </summary>
        public System.Collections.Generic.IDictionary<Model.EntityType, Model.EntityProperties> EntityProperties { get; set; }
        /// <summary>
        /// Max tick count for the game
        /// </summary>
        public int MaxTickCount { get; set; }
        /// <summary>
        /// Max pathfind nodes when performing pathfinding in the game simulator
        /// </summary>
        public int MaxPathfindNodes { get; set; }
        /// <summary>
        /// Current tick
        /// </summary>
        public int CurrentTick { get; set; }
        /// <summary>
        /// List of players
        /// </summary>
        public Model.Player[] Players { get; set; }
        /// <summary>
        /// List of entities
        /// </summary>
        public Model.Entity[] Entities { get; set; }
    
        public PlayerView(int myId, int mapSize, bool fogOfWar, System.Collections.Generic.IDictionary<Model.EntityType, Model.EntityProperties> entityProperties, int maxTickCount, int maxPathfindNodes, int currentTick, Model.Player[] players, Model.Entity[] entities)
        {
            this.MyId = myId;
            this.MapSize = mapSize;
            this.FogOfWar = fogOfWar;
            this.EntityProperties = entityProperties;
            this.MaxTickCount = maxTickCount;
            this.MaxPathfindNodes = maxPathfindNodes;
            this.CurrentTick = currentTick;
            this.Players = players;
            this.Entities = entities;
        }
    
        /// <summary> Read PlayerView from reader </summary>
        public static PlayerView ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new PlayerView();
            result.MyId = reader.ReadInt32();
            result.MapSize = reader.ReadInt32();
            result.FogOfWar = reader.ReadBoolean();
            int entityPropertiesSize = reader.ReadInt32();
            result.EntityProperties = new System.Collections.Generic.Dictionary<Model.EntityType, Model.EntityProperties>(entityPropertiesSize);
            for (int entityPropertiesIndex = 0; entityPropertiesIndex < entityPropertiesSize; entityPropertiesIndex++)
            {
                Model.EntityType entityPropertiesKey;
                Model.EntityProperties entityPropertiesValue;
                entityPropertiesKey = EntityTypeHelper.ReadFrom(reader);
                entityPropertiesValue = Model.EntityProperties.ReadFrom(reader);
                result.EntityProperties.Add(entityPropertiesKey, entityPropertiesValue);
            }
            result.MaxTickCount = reader.ReadInt32();
            result.MaxPathfindNodes = reader.ReadInt32();
            result.CurrentTick = reader.ReadInt32();
            result.Players = new Model.Player[reader.ReadInt32()];
            for (int playersIndex = 0; playersIndex < result.Players.Length; playersIndex++)
            {
                result.Players[playersIndex] = Model.Player.ReadFrom(reader);
            }
            result.Entities = new Model.Entity[reader.ReadInt32()];
            for (int entitiesIndex = 0; entitiesIndex < result.Entities.Length; entitiesIndex++)
            {
                result.Entities[entitiesIndex] = Model.Entity.ReadFrom(reader);
            }
            return result;
        }
    
        /// <summary> Write PlayerView to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(MyId);
            writer.Write(MapSize);
            writer.Write(FogOfWar);
            writer.Write(EntityProperties.Count);
            foreach (var entityPropertiesEntry in EntityProperties)
            {
                var entityPropertiesKey = entityPropertiesEntry.Key;
                var entityPropertiesValue = entityPropertiesEntry.Value;
                writer.Write((int) (entityPropertiesKey));
                entityPropertiesValue.WriteTo(writer);
            }
            writer.Write(MaxTickCount);
            writer.Write(MaxPathfindNodes);
            writer.Write(CurrentTick);
            writer.Write(Players.Length);
            foreach (var playersElement in Players)
            {
                playersElement.WriteTo(writer);
            }
            writer.Write(Entities.Length);
            foreach (var entitiesElement in Entities)
            {
                entitiesElement.WriteTo(writer);
            }
        }
    
        /// <summary> Get string representation of PlayerView </summary>
        public override string ToString() {
            string stringResult = "PlayerView { ";
            stringResult += "MyId: ";
            stringResult += MyId.ToString();
            stringResult += ", ";
            stringResult += "MapSize: ";
            stringResult += MapSize.ToString();
            stringResult += ", ";
            stringResult += "FogOfWar: ";
            stringResult += FogOfWar.ToString();
            stringResult += ", ";
            stringResult += "EntityProperties: ";
            stringResult += "{ ";
            int entityPropertiesIndex = 0;
            foreach (var entityPropertiesEntry in EntityProperties)
            {
                if (entityPropertiesIndex != 0) {
                    stringResult += ", ";
                }
                var entityPropertiesKey = entityPropertiesEntry.Key;
                stringResult += entityPropertiesKey.ToString();
                stringResult += ": ";
                var entityPropertiesValue = entityPropertiesEntry.Value;
                stringResult += entityPropertiesValue.ToString();
                entityPropertiesIndex++;
            }
            stringResult += " }";
            stringResult += ", ";
            stringResult += "MaxTickCount: ";
            stringResult += MaxTickCount.ToString();
            stringResult += ", ";
            stringResult += "MaxPathfindNodes: ";
            stringResult += MaxPathfindNodes.ToString();
            stringResult += ", ";
            stringResult += "CurrentTick: ";
            stringResult += CurrentTick.ToString();
            stringResult += ", ";
            stringResult += "Players: ";
            stringResult += "[ ";
            int playersIndex = 0;
            foreach (var playersElement in Players)
            {
                if (playersIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += playersElement.ToString();
                playersIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Entities: ";
            stringResult += "[ ";
            int entitiesIndex = 0;
            foreach (var entitiesElement in Entities)
            {
                if (entitiesIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += entitiesElement.ToString();
                entitiesIndex++;
            }
            stringResult += " ]";
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/pascal/codecraft/TcpReadWrite/Model/PlayerView.pas
unit PlayerView;

interface

uses
    Entity in 'Model/Entity.pas',
    EntityProperties in 'Model/EntityProperties.pas',
    EntityType in 'Model/EntityType.pas',
    Generics.Collections,
    Player in 'Model/Player.pas',
    Stream,
    SysUtils;

type
    // Information available to the player
    TPlayerView = class
        // Your player's ID
        myId: Int32;
        // Size of the map
        mapSize: Int32;
        // Whether fog of war is enabled
        fogOfWar: Boolean;
        // Entity properties for each entity type
        entityProperties: TDictionary<TEntityType, TEntityProperties>;
        // Max tick count for the game
        maxTickCount: Int32;
        // Max pathfind nodes when performing pathfinding in the game simulator
        maxPathfindNodes: Int32;
        // Current tick
        currentTick: Int32;
        // List of players
        players: TArray<TPlayer>;
        // List of entities
        entities: TArray<TEntity>;
        constructor Create(myId: Int32; mapSize: Int32; fogOfWar: Boolean; entityProperties: TDictionary<TEntityType, TEntityProperties>; maxTickCount: Int32; maxPathfindNodes: Int32; currentTick: Int32; players: TArray<TPlayer>; entities: TArray<TEntity>);
        // Read PlayerView from input stream
        class function ReadFrom(stream: TStream): TPlayerView; static;
        // Write PlayerView to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TPlayerView.Create(myId: Int32; mapSize: Int32; fogOfWar: Boolean; entityProperties: TDictionary<TEntityType, TEntityProperties>; maxTickCount: Int32; maxPathfindNodes: Int32; currentTick: Int32; players: TArray<TPlayer>; entities: TArray<TEntity>);
begin
    self.myId := myId;
    self.mapSize := mapSize;
    self.fogOfWar := fogOfWar;
    self.entityProperties := entityProperties;
    self.maxTickCount := maxTickCount;
    self.maxPathfindNodes := maxPathfindNodes;
    self.currentTick := currentTick;
    self.players := players;
    self.entities := entities;
end;

class function TPlayerView.ReadFrom(stream: TStream): TPlayerView;
var currentTick: Int32;
var entities: TArray<TEntity>;
var entitiesElement: TEntity;
var entitiesIndex: Int32;
var entityProperties: TDictionary<TEntityType, TEntityProperties>;
var entityPropertiesIndex: Int32;
var entityPropertiesKey: TEntityType;
var entityPropertiesSize: Int32;
var entityPropertiesValue: TEntityProperties;
var fogOfWar: Boolean;
var mapSize: Int32;
var maxPathfindNodes: Int32;
var maxTickCount: Int32;
var myId: Int32;
var players: TArray<TPlayer>;
var playersElement: TPlayer;
var playersIndex: Int32;
begin
    myId := stream.ReadInt32;
    mapSize := stream.ReadInt32;
    fogOfWar := stream.ReadBoolean;
    entityProperties := TDictionary<TEntityType, TEntityProperties>.Create;
    entityPropertiesSize := stream.ReadInt32;
    for entityPropertiesIndex := 1 to entityPropertiesSize do begin
        entityPropertiesKey := TEntityType(stream.ReadInt32);
        entityPropertiesValue := TEntityProperties.ReadFrom(stream);
        entityProperties.Add(entityPropertiesKey, entityPropertiesValue);
    end;
    maxTickCount := stream.ReadInt32;
    maxPathfindNodes := stream.ReadInt32;
    currentTick := stream.ReadInt32;
    players := TArray<TPlayer>.Create;
    SetLength(players, stream.ReadInt32);
    for playersIndex := 0 to Length(players) - 1 do begin
        playersElement := TPlayer.ReadFrom(stream);
        players[playersIndex] := playersElement;
    end;
    entities := TArray<TEntity>.Create;
    SetLength(entities, stream.ReadInt32);
    for entitiesIndex := 0 to Length(entities) - 1 do begin
        entitiesElement := TEntity.ReadFrom(stream);
        entities[entitiesIndex] := entitiesElement;
    end;
    result := TPlayerView.Create(myId, mapSize, fogOfWar, entityProperties, maxTickCount, maxPathfindNodes, currentTick, players, entities);
end;

procedure TPlayerView.WriteTo(stream: TStream);
var entitiesElement: TEntity;
var entityPropertiesKey: TEntityType;
var entityPropertiesValue: TEntityProperties;
var playersElement: TPlayer;
begin
    stream.WriteInt32(myId);
    stream.WriteInt32(mapSize);
    stream.WriteBoolean(fogOfWar);
    stream.WriteInt32(entityProperties.Count);
    for entityPropertiesKey in entityProperties.Keys do begin
        entityPropertiesValue := entityProperties.Items[entityPropertiesKey];
        stream.WriteInt32(ord(entityPropertiesKey));
        entityPropertiesValue.WriteTo(stream);
    end;
    stream.WriteInt32(maxTickCount);
    stream.WriteInt32(maxPathfindNodes);
    stream.WriteInt32(currentTick);
    stream.WriteInt32(Length(players));
    for playersElement in players do begin
        playersElement.WriteTo(stream);
    end;
    stream.WriteInt32(Length(entities));
    for entitiesElement in entities do begin
        entitiesElement.WriteTo(stream);
    end;
end;

function TPlayerView.ToString: ansistring;
var entitiesElement: TEntity;
var entitiesIndex: Int32;
var entityPropertiesFirst: Boolean;
var entityPropertiesKey: TEntityType;
var entityPropertiesKeyName: String;
var entityPropertiesValue: TEntityProperties;
var playersElement: TPlayer;
var playersIndex: Int32;
begin
    result := 'PlayerView {';
    result += 'myId=';
    result += IntToStr(myId);
    result += ', ';  
    result += 'mapSize=';
    result += IntToStr(mapSize);
    result += ', ';  
    result += 'fogOfWar=';
    result += BoolToStr(fogOfWar);
    result += ', ';  
    result += 'entityProperties=';
    result += '[';
    entityPropertiesFirst := true;
    for entityPropertiesKey in entityProperties.Keys do begin
        if not entityPropertiesFirst then
            result += ', ';
        entityPropertiesFirst := false;
        entityPropertiesValue := entityProperties.Items[entityPropertiesKey];
        WriteStr(entityPropertiesKeyName, entityPropertiesKey);
        result += entityPropertiesKeyName;;
        result += ': ';
        result += entityPropertiesValue.ToString;;
    end;
    result += ']';
    result += ', ';  
    result += 'maxTickCount=';
    result += IntToStr(maxTickCount);
    result += ', ';  
    result += 'maxPathfindNodes=';
    result += IntToStr(maxPathfindNodes);
    result += ', ';  
    result += 'currentTick=';
    result += IntToStr(currentTick);
    result += ', ';  
    result += 'players=';
    result += '[';
    for playersIndex := 0 to Length(players) - 1 do begin
        if playersIndex <> 0 then
            result += ', ';
        playersElement := players[playersIndex];
        result += playersElement.ToString;;
    end;
    result += ']';
    result += ', ';  
    result += 'entities=';
    result += '[';
    for entitiesIndex := 0 to Length(entities) - 1 do begin
        if entitiesIndex <> 0 then
            result += ', ';
        entitiesElement := entities[entitiesIndex];
        result += entitiesElement.ToString;;
    end;
    result += ']';
    result += '}';
end;

end.